src/libshadertrap/include/libshadertrap/parser.h

```diff
@@ -62,8 +62,9 @@ class Parser {
 
   bool ParseParameters(
       const std::map<Token::Type, std::function<bool()>>& parameter_parsers,
-      const std::map<Token::Type, Token::Type>& mutually_exclusive);
-
+      const std::map<Token::Type, Token::Type>& mutually_exclusive,
+      const std::map<Token::Type, std::function<bool()>>& optional_params);
```

Comment:
Ah, I see. OK - I had misread this.

I would like you to handle optional parameters a bit differently, as follows:

- Have `parameter_parsers` include *all* of the parsers - including the parsers for optional parameters
- Change `optional_params` to a `std::unordered_set<Token::Type>`, capturing those parameters that it's OK not to see during parsing.

Then the implementation of `ParseParameters` can be almost identical to what it was before your change, except that this error checking loop can be adapted to ignore optional parameters:

```
  for (const auto& entry : parameter_parsers) {
    if (already_handled.count(entry.first) == 0 &&
        observed.count(entry.first) == 0) {
      message_consumer_->Message(
          MessageConsumer::Severity::kError, tokenizer_->PeekNextToken().get(),
          "Missing parameter '" + Tokenizer::KeywordToString(entry.first) +
              "'");
      found_errors = true;
    }
  }
```

It can change to something like:

```
  for (const auto& entry : parameter_parsers) {
    if (already_handled.count(entry.first) == 0 &&
        optional_params.count(entry.first == 0) &&
        observed.count(entry.first) == 0) {
      message_consumer_->Message(
          MessageConsumer::Severity::kError, tokenizer_->PeekNextToken().get(),
          "Missing parameter '" + Tokenizer::KeywordToString(entry.first) +
              "'");
      found_errors = true;
    }
  }
```